services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/extensions/VersionedRecordExtension.java

```diff
@@ -123,9 +178,12 @@ public WriteModification beforeWrite(DynamoDbExtensionContext.BeforeWrite contex
                 throw new IllegalArgumentException("Version attribute appears to be the wrong type. N is required.");
             }
 
-            int existingVersion = Integer.parseInt(existingVersionValue.get().n());
+            long existingVersion = Long.parseLong(existingVersionValue.get().n());
             String existingVersionValueKey = VERSIONED_RECORD_EXPRESSION_VALUE_KEY_MAPPER.apply(versionAttributeKey.get());
-            newVersionValue = AttributeValue.builder().n(Integer.toString(existingVersion + 1)).build();
+
+            long increment = versionIncrementByFromAnnotation.orElse(this.incrementBy);
+            newVersionValue = AttributeValue.builder().n(Long.toString(existingVersion + increment)).build();
```

Comment:
It doesn't look like the original version handled this, but how do we deal with integer overflow? Originally `incrementBy` was always `1` so it's unlikely to happen, but with `incrementBy` being configurable now, it's technically more likely now if the user sets a relatively large value.

Can we check how v1's mapper handles this (if at all)?

Reply:
Doesn't seem like v1 had any sort of integer overflow protection? 

https://github.com/aws/aws-sdk-java/blob/c34aa7f14d7598f033091b7b9be96ee82d2c3860/aws-java-sdk-dynamodb/src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/DynamoDBMapper.java#L2454-L2470

Reply:
Thanks for looking! I'd still like us to add protection for 2.x

Reply:
Oh I see, I completely missed this. So just to double check, in order to make it backwards compatible, the new behavior should be like the following:
- if startAt is not configured: we only add version if the version attribute does not exist or is null
- if startAt is configured and the version attribute is equal to startAt: we will treat it as initial version

Is that correct?

Reply:
@zoewangg yes you are correct.

Reply:
Any updates on overflow protection? +1 to what Dongie mentioned, it's probably better to fail explicitly with a clear error message for that case

Reply:
Added overflow protection + test case.